Chap_API_Sharing_Basics.tex

```diff
@@ -0,0 +1,630 @@
+The way reserved keys are accessed by clients 
+is different than for non-reserved keys as explained in \ref{chap:api_rsvd_keys}.
+
+Host environments are not required to provide all keys that are documented in the standard.  
+However, when providing keys that are documented, they must adhere to the behavior defined by the standard for reserved keys.  
+An implementation cannot provide a reserved key that behaves differently than what is in the standard.  
+Implementations are encouraged to standardize any non-standardized, reserved keys that they provide to avoid conflicting with other implementations or efforts to standardize the same key.
+
+Host environments are permitted to provide non-reserved keys and non-standardized reserved keys.  
+However there are important caveats to providing either of these categories of keys 
+that are explained in \ref{chap:data_sharing:non_rsvd_keys} and \ref{chap:api_rsvd_keys}.
+% Would like to make this statement stronger, but there is text in chapter 6 currently that
+% says it is allowed to use "custom-prefixed keys which may adhere to either the reserved or non-reserved
+% retrieval rules.
+Host environments should avoid providing reserved keys which are not standardized.  
+If the reserved key is later standardized to behave in a way that is different from the behavior 
+provided by the host environment, the host environment will not be compliant with the PMIx standard.
+ 
+Reserved keys are set by the host environment and are made available at client initialization. 
+Each key is associated with a data realm and the retrieval API call will indicate the realm being queried.
+Reserved keys are used to access information about a client's execution context.
+After initialization, the client can access reserved keys using the PMIx_Get API.  
+Common information that a client may access includes, but is not limited to,
+the name of the host it is executing on, 
+the number of peer processes in its namespace,
+the number of peer processes in its application,
+and its job rank.
+
+%%% (Do we want to make any other statements about how and when values can change during 
+%%%  the course of execution.
+%%%  PMIX_EXIT_CODE for example, UNIVERSE_SIZE could change if more resources are added?)
+
+Non-reserved keys are provided by clients.
+A call to PMIx_Put will result in a key-value being associated with the calling client within 
+its process data realm.
+There is no mechanism to retract a key-value that has been made available by a client through \refapi{PMIx_Put}.
+The key can be assigned a new value, but cannot be removed entirely.
+Although key values can originate from different sources, they are always retrieved using the
+\refapi{PMIx_Get} and \refapi{PMIx_Get_nb} API.
+
+The publish/lookup data realm is accessed through a separate set of API's.
+\refapi{PMIx_Put} cannot add or modify key-values within the publish/lookup realm and 
```

Comment:
`PMIx_Put` has nothing to do with publish/lookup - I find this confusing mainly because of reuse of the "realm" term. You might want to reserve "realm" for the session/job/app/proc discussion to avoid confusion.

Reply:
We have talked about this.  I tried to extend the concept of data realms to all keys.  At the time, its seemed clearer to say that all keys belong to a realm and realms do not overlap.   This seems easier to present rather than saying that "keys that are accessed via Get or Query are in realms, but keys accessed by publish/lookup are not in a realm".  However, it does have the undesirable side effect that you point out that you are sometimes forced to talk about publish/lookup at times when it would be more convenient to not address those functions.   I can go either way, it's a pretty minor change to just make realms about non-published data.